{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import scipy as scp\n",
    "import re\n",
    "import os\n",
    "import sys\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Align\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "from joblib import load\n",
    "\n",
    "# adding project folder to lib search path\n",
    "project = \"/\".join((os.path.split(os.getcwd())[0], os.path.split(os.getcwd())[1], 'scripts'))\n",
    "if project not in sys.path:\n",
    "    sys.path.append(project)\n",
    "    \n",
    "from loader import *\n",
    "from work_with_files import *\n",
    "from other_functions import *\n",
    "import phylo_tree\n",
    "import classificator_k_mer\n",
    "from data_prep_k_mer import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import f1_score, accuracy_score\n",
    "from sklearn import preprocessing\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.ensemble import AdaBoostClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "haplo_seq_dict, days = work_with_files.read_fasta_haplo('data/fasta/hivevo_p2_V3.fasta')\n",
    "seq_dict, seq_name_days_dict = other_functions.preparing_data(haplo_seq_dict, days)\n",
    "prot_dict = other_functions.translating_seqs(seq_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hydro / n(SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Dealing with hydrophobicity.csv\n",
    "\n",
    "hydro_df = pd.read_csv('hydrophobicity.csv', index_col = None)\n",
    "hydro_df\n",
    "hydro_fact = dict(zip(list(hydro_df['aa']),list(hydro_df['Kyte-Doolittle'])))\n",
    "hydro_fact['X'] = 0 # !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!11  added\n",
    "hydro_fact['*'] = 0\n",
    "hydro_fact['-'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculating_snp(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Function to calculate SNP in two sequences using their alignment score\n",
    "    \n",
    "    Args:\n",
    "        seq1: string\n",
    "        seq2: string\n",
    "        \n",
    "    Return:\n",
    "        snp: int, num of SNPs\n",
    "    \"\"\"\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    score = aligner.score(seq1, seq2)\n",
    "    \n",
    "    snp = len(seq1) - score\n",
    "    \n",
    "    '''n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    if m > n:\n",
    "        n, m = m, n\n",
    "        seq1, seq2 = seq2, seq1\n",
    "    seq2 += ' '*(n-m)\n",
    "    \n",
    "    snp = 0\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            snp += 1'''\n",
    "    \n",
    "    return snp\n",
    "\n",
    "\n",
    "# calculating SNP's for every haplotype\n",
    "# snp_dict = { day: [list of SNP's for every haplo ]}\n",
    "snp_dict = {day:[] for day in days}\n",
    "ref_seq = seq_dict['reference']\n",
    "for day in seq_name_days_dict.keys():\n",
    "    for seq_name in seq_name_days_dict[day]:\n",
    "        snp_dict[day].append(calculating_snp(ref_seq, seq_dict[seq_name]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_metric(seq):\n",
    "    \"\"\"\n",
    "    Function to calculate hydrophobicity for DNA sequence\n",
    "    Args:\n",
    "        seq: string, sequnce to translate\n",
    "    Return:\n",
    "        res: float, metric result\n",
    "    \"\"\"\n",
    "    prot = translate_dna(seq, gap='-')\n",
    "    \n",
    "    res = 0\n",
    "    \n",
    "    for aa in prot:\n",
    "        res += hydro_fact[aa]\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# All this function are to make graphs\n",
    "\n",
    "def hydro_snp(seq_name_days_dict, seq_dict):\n",
    "    \"\"\" \n",
    "    Make X and Y for graph SNPs -> X, Hydro_metric -> Y\n",
    "    Remind: here only one patient is analyzed\n",
    "    Args:\n",
    "        seq_name_days_dict: dict, {day: [list of name for every haplo for this day]}\n",
    "        seq_dict: dict, {name: sequence}\n",
    "    Returns:\n",
    "        X: list, list of SNPs\n",
    "        Y: list, list of metric_\n",
    "        \"\"\"\n",
    "    \n",
    "    # making dir for SNPs; {Num of SNPs: [metric for all seq with that num of SNPs]}\n",
    "    snp_hydro_dict = {}\n",
    "    \n",
    "    # will analyze reference separetely from others\n",
    "    ref_seq = seq_dict['reference']\n",
    "    snp_hydro_dict[0] = [calculate_metric(ref_seq)]\n",
    "    \n",
    "    # for-loop for analyzing (counting metrics) for every haplo\n",
    "    # taking one day\n",
    "    for day in seq_name_days_dict.keys():\n",
    "        # taking all seqs for this day\n",
    "        for seq_name in seq_name_days_dict[day]:\n",
    "            # calculating metric\n",
    "            snp = calculating_snp(ref_seq, seq_dict[seq_name])\n",
    "            \n",
    "            # if for adding snp in dict, if it wasn't here\n",
    "            if snp not in snp_hydro_dict:\n",
    "                snp_hydro_dict[snp] = []\n",
    "            # appending to list\n",
    "            snp_hydro_dict[snp].append(calculate_metric(seq_dict[seq_name]))\n",
    "    \n",
    "    # creating lists to return\n",
    "    X = []\n",
    "    Y = []\n",
    "\n",
    "    for snp in sorted(snp_hydro_dict.keys()):\n",
    "        X.append(snp)\n",
    "        Y.append(np.array(snp_hydro_dict[snp]).mean())\n",
    "    \n",
    "    return X, Y\n",
    "\n",
    "def pat_snp(patients):\n",
    "    \"\"\"\n",
    "    Taking needed patients and using all functions above on it\n",
    "    Args:\n",
    "        patients: list with all patients\n",
    "    Return:\n",
    "        pat_X: list, with all Xs, where X is like in hydro_snp func\n",
    "        pat_Y: list, with all Ys, where X is like in hydro_snp func\n",
    "    \"\"\"\n",
    "    pat_X, pat_Y = [], []\n",
    "    \n",
    "    for patient in patients:\n",
    "        # We will not use patient#3 and patient#10 because their HIV wasn't cool at all\n",
    "        # joke, additional info can be found here (https://elifesciences.org/articles/11282)\n",
    "        if patient != 'p3' and patient != 'p10':\n",
    "            # reading fasta for patient\n",
    "            haplo_seq_dict, days = work_with_files.read_fasta('data/fasta/hivevo_{}_V3.fasta'.format(patient))\n",
    "            \n",
    "            # using all skills and mana on it\n",
    "            \n",
    "            seq_dict, seq_name_days_dict = other_functions.preparing_data(haplo_seq_dict, days)\n",
    "            prot_dict = other_functions.translating_seqs(seq_dict)\n",
    "            \n",
    "            # creating X and Y\n",
    "            X, Y = hydro_snp(seq_name_days_dict, seq_dict)\n",
    "            \n",
    "            # appending\n",
    "            pat_X.append(X)\n",
    "            pat_Y.append(Y)\n",
    "    \n",
    "    return pat_X, pat_Y\n",
    "\n",
    "pat_X, pat_Y = pat_snp(patients)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Plotting\n",
    "\n",
    "fig, axs = plt.subplots(3, 3, figsize = (35, 20), sharex = 'none', sharey = 'none')\n",
    "#fig.figsize = (20, 16)\n",
    "i = 0\n",
    "\n",
    "for x in range(3):\n",
    "    for y in range(3):\n",
    "        axs[x, y].plot(pat_X[i], pat_Y[i])\n",
    "        axs[x, y].set_title('patient {}'.format(patients_used[i]))\n",
    "        axs[x, y].grid()\n",
    "        i += 1\n",
    "\n",
    "# Sadly it should be here because of my newbie skills in pyplot\n",
    "axs[0, 0].set(ylabel='Hydrophobicity, score')\n",
    "axs[1, 0].set(ylabel='Hydrophobicity, score')\n",
    "axs[2, 0].set(ylabel='Hydrophobicity, score')\n",
    "\n",
    "axs[2, 0].set(xlabel='Num SNPs')\n",
    "axs[2, 1].set(xlabel='Num SNPs')\n",
    "axs[2, 2].set(xlabel='Num SNPs')\n",
    "\n",
    "fig.suptitle('Hydrophobicity changes due to mutations in V3 region', fontsize = 40)\n",
    "plt.savefig('graphs_hydro.png', quality=100, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using classificator (2-mers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to create graphs for 2-mers analyze\n",
    "\n",
    "def making_aa_k_mers(k):\n",
    "    \"\"\"\n",
    "    Making all possible subsequences with length k using aminoacids (order is important)\n",
    "    Args:\n",
    "        k: int, length of k-mer\n",
    "\n",
    "    Returns:\n",
    "        list of all possible aminoacid k-mer sequences\n",
    "    \"\"\"\n",
    "    global amino_string\n",
    "\n",
    "    # making all possible substrings\n",
    "\n",
    "    subseq_iter = itertools.product(amino_string, repeat=k)\n",
    "    aa_k_mer_list = list(subseq_iter)\n",
    "    del subseq_iter\n",
    "\n",
    "    # one \"for\" to deal with tuples which we get from itertools stuff\n",
    "\n",
    "    for i in range(len(aa_k_mer_list)):\n",
    "        tup = aa_k_mer_list[i]\n",
    "        aa_k_mer_list[i] = ''.join(tup)\n",
    "\n",
    "    return aa_k_mer_list\n",
    "\n",
    "\n",
    "def clf_metric_2_mer_path(path_tree, prot_dict, aa_k_mer_list):\n",
    "    \"\"\"\n",
    "    Converting sequence from path_tree into object that can be used in our classificator\n",
    "    \n",
    "    Args:\n",
    "        path_tree: list, list with names of haplotypes in exact path in tree \n",
    "        prot_dict: dict, \n",
    "        aa_k_mer_list: list, all posible k-mers from aminoacids\n",
    "    \n",
    "    Returns:\n",
    "        list of all possible aminoacid k-mer sequences\n",
    "    \"\"\"\n",
    "    metric_path = []\n",
    "    \n",
    "    # for-loop for names in path_tree\n",
    "    for name in path_tree:\n",
    "        metric_path.append([])\n",
    "        # finding prot for exact name\n",
    "        prot = prot_dict[name]\n",
    "        # finding frequency metrics\n",
    "        tmp = data_prep_k_mer.finding_freq_single_protein(prot, aa_k_mer_list)\n",
    "        \n",
    "        # KOCTblJLb (our vector from finding_freq_bla_bla is not feeling well (my bad :))\n",
    "        for x in tmp:\n",
    "            metric_path[-1].append(float(x[0]))\n",
    "        \n",
    "    return metric_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Our classificator\n",
    "\n",
    "forest_2_mer = load('saved_sklearn_models/forest.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making font BIGGER!\n",
    "\n",
    "plt.rcParams['font.size'] = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_patients(patients):\n",
    "    \"\"\"\n",
    "    Function to analyze all paths in created Phylo_tree for each patient. \n",
    "    Here probability to be human's protein is stored in Y, which then will be plotted. \n",
    "    Args:\n",
    "        patients: list, list of all patients\n",
    "    Return:\n",
    "        pat_X: list, list of Xs, X = [days, when patient was sequenced]\n",
    "        pat_Y: list, list of Ys, Y = [probability in particular day]; X and Y are connected (have same lengths)\n",
    "    \"\"\"\n",
    "    \n",
    "    # Our lists\n",
    "    pat_X = []\n",
    "    pat_Y = []\n",
    "    \n",
    "    aa_k_mer_list = making_aa_k_mers(2)\n",
    "    \n",
    "    # for-loop for patients\n",
    "    for patient in patients:\n",
    "\n",
    "        # We will not use patient#3 and patient#10 because their HIV wasn't cool at all\n",
    "        # joke, additional info can be found here (https://elifesciences.org/articles/11282)\n",
    "        \n",
    "        if patient != 'p3' and patient != 'p10':\n",
    "            \n",
    "            # reading fasta\n",
    "            haplo_seq_dict, days = work_with_files.read_fasta('data/fasta/hivevo_{}_V3.fasta'.format(patient))\n",
    "            \n",
    "            # using all our AOE-skills and mana on it\n",
    "            \n",
    "            seq_dict, seq_name_days_dict = other_functions.preparing_data(haplo_seq_dict, days)\n",
    "            prot_dict = other_functions.translating_seqs(seq_dict)\n",
    "\n",
    "            # Constructing tree\n",
    "            phylo_tree = phylo_tree.Phylo_tree()\n",
    "            phylo_tree.construct_levels(days)\n",
    "            phylo_tree.constructing_tree(seq_dict, seq_name_days_dict)\n",
    "            phylo_paths = phylo_tree.create_all_path()\n",
    "            \n",
    "            # Making X\n",
    "            X = list(map(int, days))\n",
    "            X.insert(0, 0)\n",
    "            pat_X.append(X)\n",
    "            pat_Y.append([])\n",
    "            \n",
    "            # Using classificator to find out probability to be human's gene\n",
    "            for path in phylo_paths:\n",
    "                met = clf_metric_2_mer_path(path, prot_dict, aa_k_mer_list)\n",
    "                Y = forest_2_mer.predict_proba(met)[:, 1]\n",
    "                pat_Y[-1].append(Y)\n",
    "            \n",
    "    return pat_X, pat_Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting\n",
    "\n",
    "fig, axs = plt.subplots(3, 3, figsize = (35, 20), sharex = 'none', sharey = 'none')\n",
    "#fig.figsize = (20, 16)\n",
    "i = 0\n",
    "for x in range(3):\n",
    "    for y in range(3):\n",
    "        for Y in pat_Y[i]:\n",
    "            axs[x, y].plot(pat_X[i], Y)\n",
    "            axs[x, y].set_title('patient {}'.format(patients_used[i]))\n",
    "        axs[x, y].grid(alpha=1)\n",
    "        i += 1\n",
    "\n",
    "# Sadly it should be here because of my newbie skills in pyplot\n",
    "axs[0, 0].set(ylabel='Prob to be human prot, %')\n",
    "axs[1, 0].set(ylabel='Prob to be human prot, %')\n",
    "axs[2, 0].set(ylabel='Prob to be human prot, %')\n",
    "\n",
    "axs[2, 0].set(xlabel='time, days')\n",
    "axs[2, 1].set(xlabel='time, days')\n",
    "axs[2, 2].set(xlabel='time, days')\n",
    "\n",
    "fig.suptitle('Evolution of HIV of V3 genome region', fontsize = 40)\n",
    "plt.savefig('graphs.png', quality=100, bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
